include/circt/Dialect/SMT/SMTAttributes.td

```diff
@@ -0,0 +1,61 @@
+//===- SMTAttributes.td - Attributes for SMT dialect -------*- tablegen -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines SMT dialect specific attributes.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef CIRCT_DIALECT_SMT_SMTATTRIBUTES_TD
+#define CIRCT_DIALECT_SMT_SMTATTRIBUTES_TD
+
+include "circt/Dialect/SMT/SMTDialect.td"
+include "mlir/IR/EnumAttr.td"
+include "mlir/IR/BuiltinAttributeInterfaces.td"
+
+def BitVectorAttr : AttrDef<SMTDialect, "BitVector", [
+  DeclareAttrInterfaceMethods<TypedAttrInterface>
+]> {
+  let mnemonic = "bv";
+  let description = [{
+    This attribute represents a constant value of the `(_ BitVec width)` sort as
+    described in the [SMT bit-vector
+    theory](https://smtlib.cs.uiowa.edu/theories-FixedSizeBitVectors.shtml).
+
+    The constant is parsed and printed as #bX (binary) or #xX (hexadecimal)
+    where X is the value in the corresponding format without any further
+    prefixing just as in SMT-LIB. The number of digits determines the bit-width
+    of the bit-vector. This means, leading zeros are important!
+
+    Examples:
+    ```mlir
+    #smt.bv<"#b0101"> : !smt.bv<4>
+    #smt.bv<"#x5c"> : !smt.bv<8>
+    ```
```

Comment:
I think I just went for this syntax for the more SMT-like feel, but using regular decimal numbers is also a good approach. It's probably just a matter of taste. The only disadvantages of using regular decimals I see are
* bit-vector width cannot be inferred from the constant, i.e., we need the explicit type annotation: `smt.bv.constant #smt.bv<5> : !smt.bv<8>` vs `smt.bv.constant "#x05"` which basically means we always need the `quantified` version of the attribute in any operation assembly formats, on the other hand it saves typing a lot of zeros for big bit-vectors just storing small numbers
* need to do some APInt `zext` and `trunc` with sanity checks, i.e., we just replace the complexity with some other complexity
* I don't really understand why MLIR doesn't allow integer constants of the format `0b1010` which is the most basic way of representing bit-vectors

Reply:
I might disagree with the first point, because while `#x05` is easy to read and to understand, I cannot tell you from a glance what`s the bitwidth of `#b010101010101101100011101010`, so having an explicit bitwidth makes a lot of sense to me!
What is the difference with `zext` and `trunc`? My understanding is that this would be the same? At least, I do not recommend changing the storage of `#smt.bv`.

And I would agree for the third point, though the problem is knowing when to use that format, and when to use the decimal format.

Reply:
When parsing an APInt, the bitwidth does not necessarily match the one required for the given bitvector type and has to be adjusted in the attribute parser with `sext` and `trunc` and before truncating it should be checked if the truncated bits are significant/meaningful. The storage of `#smt.bv` is the autogenerated one with just a single `APInt` since that stores both the value and the bitwidth.